src/libostree/ostree-repo.c

```diff
@@ -347,49 +347,56 @@ push_repo_lock (OstreeRepo          *self,
     {
       g_debug ("Creating repo lock table");
       lock_table = g_hash_table_new_full (NULL, NULL, NULL,
-                                          (GDestroyNotify)free_repo_lock);
+                                          (GDestroyNotify)g_ptr_array_unref);
       g_private_set (&repo_lock_table, lock_table);
     }
 
-  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
-  if (lock == NULL)
+  GPtrArray *stack = g_hash_table_lookup (lock_table, self);
+  if (stack == NULL)
+    {
+      stack = g_ptr_array_new ();
+      g_hash_table_insert (lock_table, self, stack);
+    }
+
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&self->lock.mutex);
+
+  if (self->lock.fd == -1)
     {
-      lock = g_new0 (OstreeRepoLock, 1);
-      g_queue_init (&lock->stack);
       g_debug ("Opening repo lock file");
-      lock->fd = TEMP_FAILURE_RETRY (openat (self->repo_dir_fd, ".lock",
-                                             O_CREAT | O_RDWR | O_CLOEXEC,
-                                             DEFAULT_REGFILE_MODE));
-      if (lock->fd < 0)
-        {
-          free_repo_lock (lock);
-          return glnx_throw_errno_prefix (error,
-                                          "Opening lock file %s/.lock failed",
-                                          gs_file_get_path_cached (self->repodir));
-        }
-      g_hash_table_insert (lock_table, self, lock);
+      self->lock.fd = TEMP_FAILURE_RETRY (openat (self->repo_dir_fd, ".lock",
+                                                  O_CREAT | O_RDWR | O_CLOEXEC,
+                                                  DEFAULT_REGFILE_MODE));
+      if (self->lock.fd < 0)
+        return glnx_throw_errno_prefix (error,
+                                        "Opening lock file %s/.lock failed",
+                                        gs_file_get_path_cached (self->repodir));
     }
 
   OstreeRepoLockInfo info;
-  repo_lock_info (lock, &info);
+  repo_lock_info (self, &info);
   g_debug ("Push lock: state=%s, depth=%u", info.name, info.len);
 
   if (info.state == LOCK_EX)
     {
       g_debug ("Repo already locked exclusively, extending stack");
-      g_queue_push_head (&lock->stack, GINT_TO_POINTER (LOCK_EX));
+      self->lock.exclusive++;
```

Comment:
I think this can overflow (wrap back to 0), breaking lock correctness.

Reply:
I'm inclined to think that it should fail if you're going to overflow since I can't imagine any real application intentionally recursing the lock that far. What do you think?

Reply:
Yeah, we could add `g_assert_cmpint (self->lock.exclusive, <, G_MAXINT);` but it seems unlikely.
Ah, TIL GCC [has builtins for this](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html#Integer-Overflow-Builtins) similar to what Rust exposes via e.g. [https://doc.rust-lang.org/std/primitive.i32.html#method.overflowing_add](overflowing_add).

Reply:
glib even has wrappers for the GCC builtins in https://developer.gnome.org/glib/stable/glib-Bounds-checked-integer-arithmetic.html. These were added in 2.48. However, after looking at adding the assertions, I'm pretty sure I'd do the check and increment separately. The check would be before changing the file lock state, and the increment would be after the file lock call succeeded.